src/databricks/labs/ucx/install.py

```diff
@@ -676,17 +675,39 @@ def _get_workspace(workspace_id: int, ids_to_workspace: dict[int, Workspace]) ->
             msg = "Current workspace is not known, Please run as account-admin: databricks labs ucx sync-workspace-info"
             raise KeyError(msg) from None
 
-    def _get_workspace_info(self, current_workspace_id: int):
+    def _get_workspace_ids(self):
         account_client = self._get_safe_account_client()
-        workspace = account_client.workspaces.get(current_workspace_id)
-        current_workspace_client = account_client.get_workspace_client(workspace)
-        installation = Installation.current(current_workspace_client, self.product_info.product_name())
-        workspace_info = WorkspaceInfo(installation, current_workspace_client)
-        return workspace_info.load_workspace_info()
+        workspaces = account_client.workspaces.list()
+        ids_to_workspace = {}
+        for workspace in workspaces:
+            ids_to_workspace[workspace.workspace_id] = workspace
+        return ids_to_workspace
+
+    def get_workspace_contexts(self, collection_workspace_id: int) -> list[WorkspaceContext]:
+        workspace_contexts = []
+        account_client = self._get_safe_account_client()
+        acct_ctx = AccountContext(account_client)
+        collection_workspace = account_client.workspaces.get(collection_workspace_id)
+        if not acct_ctx.account_workspaces.can_administer(collection_workspace):
+            logger.error(f"User is not workspace admin of collection workspace {collection_workspace_id}")
+            return []
+        installer = self._get_installer(collection_workspace)
+        if installer.config.installed_workspace_ids is None:
+            logger.error(f"No collection info found in the workspace {collection_workspace_id}")
+            return []
+        for workspace_id in installer.config.installed_workspace_ids:
+            workspace = account_client.workspaces.get(workspace_id)
+            if not acct_ctx.account_workspaces.can_administer(workspace):
+                logger.error(f"User is not workspace admin of workspace {workspace_id}")
+                return []
+            workspace_client = account_client.get_workspace_client(workspace)
+            ctx = WorkspaceContext(workspace_client)
+            workspace_contexts.append(ctx)
+        return workspace_contexts
 
     def join_collection(
         self,
-        current_workspace_id: int,
+        workspace_ids: list[int],
         target_workspace_id: int | None = None,
```

Comment:
signature is a bit strange - if we have `target_workspace_id` and `workspace_ids` - what's the difference?... how would `ctx.account_workspaces.get_accessible_workspaces()` semantics change?

Reply:
So join collection is invoked from cli command. You can supply a list of workspace id in the workspace_ids variable which you want to join to a collection. target_workspace_id is the workspace to which's collection you want to join.
Ex if target_workspace_id (111) has a collection with workspaces ids 111,222,333. You want to join new set of workspace id to the collection (555,666) Then you pass [555,666] to workspace_ids and 111 to target_workspace_id and the new collection will be [111,222,333,444,555].

The above works when you are passing a target_workspace_id. If you are not passing, then code will run the get_accessible_workspace and get the list of eligible workspace(which you are admin on) which you can join as a collection.

Reply:
> target_workspace_id is the workspace to which's collection you want to join.

do we need a "target workspace"? isn't the supplied list of workspaces a target? the point of this command is to make sure that all workspaces in a collection know about other workspaces. effectively modifying IDs everywhere.

isn't the intended flow as:

1. get list of every other workspaces
2. for every other workspace, get a list of what they know of other workspaces
3. combine all these workspace ids
4. for every workspace id, modify configuration of UCX and put the same workspace ids everywhere

Reply:
ok. so in my example above, workspace 555 and 666 are not part of any collection. workspaces 111,222,333 are part of a collection. so we run a command saying join collection with workspace_ids as [111,555,666] and resultant will be all 5 workspaces will be part of the collection.?

Reply:
> resultant will be all 5 workspaces will be part of the collection

yes, isn't that the goal? the less ambiguous arguments the function has - the better.

Reply:
Updated the logic to reflect this. when run from cli it will pass the list of workspace id. Run doing it as part of the ucx installation pass a boolean value it tell users to prompt for joining collection and also lists the existing installation to join as collection